docker/driver.go

```diff
@@ -0,0 +1,105 @@
+package docker
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"github.com/bblfsh/performance"
+
+	"bitbucket.org/creachadair/shell"
+	"github.com/creachadair/staticfile"
+	"gopkg.in/src-d/go-errors.v1"
+	"gopkg.in/src-d/go-log.v1"
+)
+
+//go:generate go run github.com/creachadair/staticfile/compiledata -pkg docker -out static.go conf/drivers.json
+
+// InstallDriver downloads driver's repository, checkouts to a given commit,
+// builds docker image and installs driver's image to bblfshd container.
+// Requires bblfshd container running.
+func InstallDriver(language, commit string) (func(), error) {
+	data, err := staticfile.ReadAll("conf/drivers.json")
+	if err != nil {
+		return func() {}, errors.NewKind("failed to read drivers config").Wrap(err)
+	}
+
+	drivers := make(map[string]string)
+	if err := json.Unmarshal(data, &drivers); err != nil {
+		return nil, err
+	}
+	log.Debugf("Supported drivers: %v", drivers)
+
+	driver, ok := drivers[language]
+	if !ok {
+		return func() {}, errors.NewKind("language %v is not supported").New(language)
+	}
+	log.Debugf("Selected driver: %v", driver)
+
+	image := driverImage(driver, commit)
+	cleanup, err := downloadAndBuildDriver(driver, commit, image)
+	if err != nil {
+		return cleanup, err
+	}
+	if err := installDriverToBblfshd(bblfshdContainer, language, image); err != nil {
+		return cleanup, err
+	}
+
+	return cleanup, err
+}
+
+// downloadAndBuildDriver creates directory in the temporary folder, clones driver's repository there, checkouts to a given commit
+// and runs docker image build script.
+// Arguments:
+// driver - name of the supported driver(check docker/conf/drivers.json)
+// commit - commit hash to checkout to
+// image - desired docker image name as the result of the build
+// Returns: cleanup function and error
+func downloadAndBuildDriver(driver, commit, image string) (func(), error) {
+	tmpDir, err := ioutil.TempDir("", driver)
+	if err != nil {
+		return func() {}, err
+	}
+	log.Debugf("Created temp directory %v", tmpDir)
+
+	cleanup := func() { os.RemoveAll(tmpDir) }
+
+	repo := driverURL(driver)
+	log.Debugf("performing git clone repository %v %v", repo, commit)
+	cloneCommand := fmt.Sprintf("git clone %[1]s %[2]s ; cd %[2]s ; git checkout %[3]s",
+		shell.Quote(repo), shell.Quote(tmpDir), shell.Quote(commit))
+	if err := performance.ExecCmd(cloneCommand); err != nil {
+		cleanup()
```

Comment:
If we expect the caller to always invoke the cleanup, we don't need to do it here. On the other hand, if we don't want the caller to deal with the error case, we should return `nil, err` instead. 

I think the API is nicer if the caller doesn't have to worry about cleanups on the error path, but I don't feel that strongly about which we do as long as we are consistent.

Reply:
If we consider the current case from the narrow perspective, we don't need this path any further as soon as docker image was build, so we don't need to worry about cleaning this path on higher levels. Please check out my solution.

Reply:
I like it, thanks!